torchao/prototype/quantization/codebook/codebook_quantized_tensor.py

```diff
@@ -0,0 +1,284 @@
+    Codebook quantized tensor subclass.
+
+    Codebook (lookup table) quantization involves partitioning the input tensor into blocks, and replacing each block
+    with the index of the closest entry in a predefined codebook.
+
+    Fields:
+      codes (torch.Tensor): Tensor of indices representing blocks in the original tensor. Each index
+         maps to a corresponding codebook entry.
+      codebook (torch.Tensor): Tensor representing the quantization codebook, where each entry
+         corresponds to a block in the original tensor. Shape is `(codebook_size, out_block_size, in_block_size)`.
+      block_size (Tuple[int, ...]): Granularity of quantization, specifying the dimensions of tensor
+         blocks that share the same quantization parameters.
+      shape (torch.Size): Shape of the original high-precision tensor.
+      dtype (torch.dtype): dtype of the original high-precision tensor.
+    """
+
+    @staticmethod
+    def __new__(
+        cls,
+        codes: torch.Tensor,
+        codebook: torch.Tensor,
+        block_size: Tuple[int, ...],
+        scales: torch.Tensor,
+        shape: torch.Size,
+        dtype=None,
+        strides=None,
+    ):
+        kwargs = {}
+        kwargs["device"] = codes.device
+        kwargs["layout"] = (
+            kwargs.get("layout") if kwargs.get("layout", False) else codes.layout
+        )
+        kwargs["dtype"] = dtype
+        if strides is not None:
+            kwargs["strides"] = strides
+        kwargs["requires_grad"] = False
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        codes: torch.Tensor,
+        codebook: torch.Tensor,
+        block_size: Tuple[int, ...],
+        scales: torch.Tensor,
+        shape: torch.Size,
+        dtype=None,
+        strides=None,
+    ):
+        self.codes = codes
+        self.codebook = codebook
+        self.block_size = block_size
+        self.scales = scales
+        self._dtype = dtype  # not sure this is right
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}(codes={self.codes}, codebook={self.codebook}, block_size={self.block_size}, scales={self.scales}, "
+            f"shape={self.shape}, device={self.device}, dtype={self.dtype}, requires_grad={self.requires_grad})"
+        )
+
+    def _quantization_type(self):
+        return f"shape={self.shape}, block_size={self.block_size}, codebook_size={self.codebook.size(0)}, device={self.device}, code_dtype={self.codes.dtype}"
+
+    def dequantize(self, output_dtype: Optional[torch.dtype] = None) -> torch.Tensor:
+        if output_dtype is None:
+            output_dtype = self.dtype
+
+        if isinstance(self.codes, UintxTensor):
+            codes = self.codes.get_plain()
+        else:
+            codes = self.codes
+        if codes.dtype == torch.uint8:
+            codes = codes.to(torch.int32)  # not sure how to index with uint8
```

Comment:
is this needed? this might be the reason why it's slow

Reply:
I think so. I do indexing in `dequant = codebook[codes]` in `dequantize_codebook`. I got an error when I when I tried doing `codebook[codes]` when codes was uint8.

Reply:
what is the error? might be easy to support this in pytorch I feel

Reply:
only int and long are supported right now: https://github.com/pytorch/pytorch/blob/6e203ae6deaceb370e497bd50f2d02e894f5e9cc/aten/src/ATen/Dispatch.h#L801

Reply:
UserWarning: indexing with dtype torch.uint8 is now deprecated, please use a dtype torch.bool instead.
IndexError: The shape of the mask [2] at index 0 does not match the shape of the indexed tensor [3, 3] at index 0.

Reply:
OK, let's add a TODO here to follow up and land this for now, is perplexity number expected here? it looks like it's slightly worse than int4wo: https://github.com/pytorch/ao/blob/main/torchao/quantization/README.md#cuda-backend

Reply:
I think perplexity is expected to be slightly worse than asymmetric int4wo (with scales and zero points) because I only add scales. But it should usually be better than symmetric int4wo (only scales, no zero points), though I initialize the centroids in kmeans randomly so it could be worse if centroids have bad initialization.